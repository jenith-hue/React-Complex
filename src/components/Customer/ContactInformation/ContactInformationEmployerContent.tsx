import {
  RACButton,
  RACTable,
  RACTableCell,
  RACTableRow,
  RACTooltip,
  Typography,
} from '@rentacenter/racstrap';
import React from 'react';
import { useLocation } from 'react-router-dom';

import { useCustomerDetails } from '../../../context/CustomerDetails/CustomerDetailsProvider';
import { CustomerLocationState } from '../../../types/types';
import { abbrev, formatPhoneNumber } from '../../../utils/utils';
import { withContactStyles, WrappedComponentProps } from './withContactStyles';

export const ContactInformationEmployerContent_ = ({
  classes,
  onLogClicked,
}: WrappedComponentProps) => {
  const {
    customerDetails: { employerReferences: data },
    isLogActivityAllowed,
  } = useCustomerDetails();
  const location = useLocation<CustomerLocationState>();

  const renderTableHead = () => (
    <>
      <RACTableCell>Company Name</RACTableCell>
      <RACTableCell>Employer Phone Number</RACTableCell>
      <RACTableCell># Contacts Today</RACTableCell>
      <RACTableCell>Last Call Result</RACTableCell>
      <RACTableCell align="center">Action</RACTableCell>
    </>
  );

  const customerId =
    location?.state?.customer?.customerId || location?.pathname?.split('/')[3];

  const renderTableContent = () => (
    <>
      {data?.map((dataObj) => (
        <RACTableRow backgroundColor="white" key={dataObj.employerReferenceId}>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {dataObj.employerName}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography variant="body2" className={classes.tableContentColor}>
              {formatPhoneNumber(
                dataObj.employerPhoneNumber,
                dataObj.doNotCall === 'Y' || !isLogActivityAllowed
              )}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography
              align="center"
              variant="body2"
              className={classes.tableContentColor}
            >
              {dataObj.communicationsToday}
            </Typography>
          </RACTableCell>
          <RACTableCell>
            <Typography noWrap variant="body2" color="primary">
              <RACTooltip title={dataObj.lastCallResultDescription || ''}>
                <span>{abbrev(dataObj.lastCallResultDescription || '')}</span>
              </RACTooltip>
            </Typography>
          </RACTableCell>
          <RACTableCell className={classes.center}>
            <RACButton
              variant="outlined"
              color="primary"
              className={classes.logButton}
              onClick={() =>
                onLogClicked(customerId, dataObj.employerPhoneNumber)
              }
              disabled={
                !dataObj?.employerPhoneNumber ||
                !isLogActivityAllowed ||
                dataObj.doNotCall === 'Y'
              }
            >
              Log
            </RACButton>
          </RACTableCell>
        </RACTableRow>
      ))}
    </>
  );
  return (
    <RACTable
      renderTableHead={renderTableHead}
      renderTableContent={renderTableContent}
    />
  );
};

export const ContactInformationEmployerContent = (props: any) =>
  withContactStyles(props)(ContactInformationEmployerContent_);
