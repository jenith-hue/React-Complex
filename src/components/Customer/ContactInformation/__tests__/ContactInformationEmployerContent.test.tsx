/* eslint-disable react/display-name */
/* eslint-disable sonarjs/no-identical-functions */

import React from 'react';
import renderer from 'react-test-renderer';
import { ContactInformationEmployerContent } from '../ContactInformationEmployerContent';
import RouteData from 'react-router-dom';

interface PropWithChildren {
  children: JSX.Element;
}

interface MockRACTableProps {
  renderTableHead: () => JSX.Element;
  renderTableContent: () => JSX.Element;
}

jest.mock('@rentacenter/racstrap', () => ({
  ...jest.requireActual('@rentacenter/racstrap'),
  Typography: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="Typography" {...rest}>
      {children}
    </div>
  ),
  RACButton: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACButton" {...rest}>
      {children}
    </div>
  ),
  RACCard: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACCard" {...rest}>
      {children}
    </div>
  ),
  RACTableCell: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableCell" {...rest}>
      {children}
    </div>
  ),
  RACTableRow: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTableRow" {...rest}>
      {children}
    </div>
  ),
  RACTable: ({ renderTableHead, renderTableContent }: MockRACTableProps) => (
    <div data-testid="RACTable">
      <div id="table-head">{renderTableHead()}</div>
      <div id="table-content">{renderTableContent()}</div>
    </div>
  ),
  RACTooltip: ({ children, ...rest }: PropWithChildren) => (
    <div data-testid="RACTooltip" {...rest}>
      {children}
    </div>
  ),
  makeStyles: () => () => ({}),
}));

jest.mock(
  '../../../../context/CustomerDetails/CustomerDetailsProvider',
  () => ({
    useCustomerDetails: () => ({
      customerDetails: {
        employerReferences: [
          {
            employerName: 'Employer name',
            employerPhoneNumber: '1234567890',
            communicationsToday: 123,
            lastCallResultDescription: 'Last call result description',
          },
        ],
      },
    }),
  })
);

// Notice that the below is required, even if there is an another
// jest.spyOn in the test.
// Otherwise, we'll get the following: TypeError: Cannot redefine property: useLocation
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '',
  }),
}));
describe('ContanctInformationEmployerContent', () => {
  it('renders correctly', () => {
    jest.spyOn(RouteData, 'useLocation').mockImplementationOnce(() => ({
      pathname: '/testroute',
      search: '',
      hash: '',
      state: {
        customer: {
          customerFirstName: 'First',
          customerLastName: 'Last',
        },
      },
    }));

    const tree = renderer.create(<ContactInformationEmployerContent />);

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
