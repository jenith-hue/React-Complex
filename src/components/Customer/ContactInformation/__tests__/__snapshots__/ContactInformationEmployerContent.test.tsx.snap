// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContanctInformationEmployerContent renders correctly 1`] = `
<div
  data-testid="RACTable"
>
  <div
    id="table-head"
  >
    <div
      data-testid="RACTableCell"
    >
      Company Name
    </div>
    <div
      data-testid="RACTableCell"
    >
      Employer Phone Number
    </div>
    <div
      data-testid="RACTableCell"
    >
      # Contacts Today
    </div>
    <div
      data-testid="RACTableCell"
    >
      Last Call Result
    </div>
    <div
      align="center"
      data-testid="RACTableCell"
    >
      Action
    </div>
  </div>
  <div
    id="table-content"
  >
    <div
      backgroundColor="white"
      data-testid="RACTableRow"
    >
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          Employer name
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          data-testid="Typography"
          variant="body2"
        >
          (***) ***-****
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          align="center"
          data-testid="Typography"
          variant="body2"
        >
          123
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="Typography"
          noWrap={true}
          variant="body2"
        >
          <div
            data-testid="RACTooltip"
            title="Last call result description"
          >
            <span>
              Last call res...
            </span>
          </div>
        </div>
      </div>
      <div
        data-testid="RACTableCell"
      >
        <div
          color="primary"
          data-testid="RACButton"
          disabled={true}
          onClick={[Function]}
          variant="outlined"
        >
          Log
        </div>
      </div>
    </div>
  </div>
</div>
`;
